pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java

```diff
@@ -0,0 +1,245 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.api;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link ConsumerBuilder} is used to configure and create instances of {@link Consumer}.
+ *
+ * @see PulsarClient#newConsumer()
+ *
+ * @since 2.0.0
+ */
+public interface ConsumerBuilder extends Serializable, Cloneable {
+
+    /**
+     * Create a copy of the current consumer builder.
+     * <p>
+     * Cloning the builder can be used to share an incomplete configuration and specialize it multiple times. For
+     * example:
+     *
+     * <pre>
+     * ConsumerBuilder builder = client.newConsumer() //
+     *         .subscriptionName("my-subscription-name") //
+     *         .subscriptionType(SubscriptionType.Shared) //
+     *         .receiverQueueSize(10);
+     *
+     * Consumer consumer1 = builder.clone().topic(TOPIC_1).subscribe();
+     * Consumer consumer2 = builder.clone().topic(TOPIC_2).subscribe();
+     * </pre>
+     */
+    ConsumerBuilder clone();
+
+    /**
+     * Finalize the {@link Consumer} creation by subscribing to the topic.
+     *
+     * <p>
+     * If the subscription does not exist, a new subscription will be created and all messages published after the
+     * creation will be retained until acknowledged, even if the consumer is not connected.
+     *
+     * @return the {@link Consumer} instance
+     * @throws PulsarClientException
+     *             if the the subscribe operation fails
+     */
+    Consumer subscribe() throws PulsarClientException;
+
+    /**
+     * Finalize the {@link Consumer} creation by subscribing to the topic in asynchronous mode.
+     *
+     * <p>
+     * If the subscription does not exist, a new subscription will be created and all messages published after the
+     * creation will be retained until acknowledged, even if the consumer is not connected.
+     *
+     * @return a future that will yield a {@link Consumer} instance
+     * @throws PulsarClientException
+     *             if the the subscribe operation fails
+     */
+    CompletableFuture<Consumer> subscribeAsync();
+
+    /**
+     * Specify the topic this consumer will subscribe on.
+     * <p>
+     * This argument is required when constructing the consumer.
+     *
+     * @param topicName
+     */
+    ConsumerBuilder topic(String topicName);
```

Comment:
It's just an idea but we may want to pass a Topic object instead of its string representation. Then this interface would be freed to define a format of topic name string (persistent://.../.../...), and users can manage their topic names in any format.

Reply:
I would leave it as a string, at least for now, because that would be a bigger change I think. We're already trying to address the long topic names by having some degree of implicitenes. See https://github.com/apache/incubator-pulsar/wiki/PIP-10:-Remove-cluster-for-namespace-and-topic-names and https://github.com/apache/incubator-pulsar/wiki/PIP-11:-Short-topic-names

Reply:
Alright. Yeah, and the change is not necessary for introducing builders.

Below is what I thought.

We should probably introduce a parser for topic name string to reduce duplicate parsing code. Otherwise PIP-10 and PIP-11 would be tough because we need to modify all parsing code here and there. If we introduce the parser, the builder implementation would be like:

```java
public ConsumerBuilder topic(String topicName) {
  this.topic = Topic.parse(topicName);
  return this;
}
```
The Topic class would be like this:
```java
class Topic {
  public Topic(String topicName) {
    // parse
    ...
    this.namespace = ....
    this.cluster = ...
    ...
  } 
  public getNamespace() {
    return this.namespace;
  }
  ...
}
```

Once we did this, other components that use Topic object wouldn't need to care how the object was created. And if so, it might be useful if users could create own Topic implementation and pass it to builders.


So I'm not saying we should eliminate the Pulsar topic URL format. I just thought it may increase its capability.

Reply:
For parsing, it's already all concentrated in the `DestinationName` class, which I plan to rename into `TopicName`. It is not exposed to users though.